processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java

```diff
@@ -651,9 +650,7 @@ public ResultRow apply(Object input)
             );
 
             if (isResultLevelCache) {
-              Iterator<PostAggregator> postItr = query.getPostAggregatorSpecs().iterator();
-              int postPos = 0;
-              while (postItr.hasNext() && results.hasNext()) {
+              for (int postPos = 0; postPos < query.getPostAggregatorSpecs().size(); postPos++) {
                 resultRow.set(postAggregatorStart + postPos, results.next());
               }
```

Comment:
I think the contract around here should be to restore the same row that it have saved into the cache - if its not able to do so; I think it should fail with an exception - if it doesn't do that wouldn't that cause incorrect results - or not?

Reply:
I was also going through this area for a different purpose, and indeed there's a check below this code that checks if both the iterators have been exhausted completely, so even with the above check, the code would fail below. However, the error message would be more descriptive, so I think we should add the check back (otherwise it would fail with a generic `NoSuchElementException` exception)

Reply:
not sure about the case you were thinking - but I'm afraid that check might not be that usefull for these cases; consider the following:
* lets have
   * a `results` with  `{1 dim}+{1 aggs}+{1 postagg} = 3`
   * a `resultRow` with `{1 dim}+{1 aggs}+{2 postagg} = 4` (for whatever reason ?)
   * since in these cases the system is already facing a serious error - I'm not sure what value a nicer error message will give - as such error could not be fixed easily for sure
* I think a more reasonable check would be to ensure that `results.size() == resultRow.size()` - so that we can remove the checking of all these iterators - as they are trying to do the same in a more complicated way

I've added an `if` to throw an exception in case there are no next when there should be one; but all these conditionals are kinda redundant as the iterator was created for a list

Reply:
https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java#L660
I was referring to this check. It should also handle the case that you mentioned above right? Perhaps adding the size check that you mentioned earlier is more legible.